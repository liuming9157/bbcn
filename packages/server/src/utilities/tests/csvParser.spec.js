const { readFileSync } = require("../fileSystem")
const csvParser = require("../csvParser")

const CSV_PATH = __dirname + "/test.csv"

const SCHEMAS = {
  VALID: {
    Age: {
      type: "number",
    },
  },
  INVALID: {
    Address: {
      type: "number",
    },
    Age: {
      type: "number",
    },
  },
  IGNORE: {
    Address: {
      type: "omit",
    },
    Age: {
      type: "omit",
    },
    Name: {
      type: "string",
    },
  },
  BROKEN: {
    Address: {
      type: "datetime",
    },
  },
}

describe("CSV Parser", () => {
  const csvString = readFileSync(CSV_PATH, "utf8")

  describe("parsing", () => {
    it("returns status and types for a valid CSV transformation", async () => {
      expect(await csvParser.parse(csvString, SCHEMAS.VALID)).toEqual({
        Address: {
          success: true,
          type: "string",
        },
        Age: {
          success: true,
          type: "number",
        },
        Name: {
          success: true,
          type: "string",
        },
      })
    })

    it("returns status and types for an invalid CSV transformation", async () => {
      expect(await csvParser.parse(csvString, SCHEMAS.INVALID)).toEqual({
        Address: {
          success: false,
          type: "number",
        },
        Age: {
          success: true,
          type: "number",
        },
        Name: {
          success: true,
          type: "string",
        },
      })
    })
  })

  describe("transformation", () => {
    it("transforms a CSV file into JSON", async () => {
      expect(
        await csvParser.transform({
          schema: SCHEMAS.VALID,
          csvString,
        })
      ).toMatchSnapshot()
    })

    it("transforms a CSV file into JSON ignoring certain fields", async () => {
      expect(
        await csvParser.transform({
          schema: SCHEMAS.IGNORE,
          csvString,
        })
      ).toEqual([
        {
          Name: "BertÃ¥",
        },
        {
          Name: "Ernie",
        },
        {
          Name: "Big Bird",
        },
      ])
    })

    it("throws an error on invalid schema", async () => {
      await expect(
        csvParser.transform({ schema: SCHEMAS.BROKEN, csvString })
      ).rejects.toThrow()
    })
  })
})
