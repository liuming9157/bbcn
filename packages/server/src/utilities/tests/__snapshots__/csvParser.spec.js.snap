// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSV Parser transformation transforms a CSV file into JSON 1`] = `
Array [
  Object {
    "Age": 4324,
  },
  Object {
    "Age": 34,
  },
  Object {
    "Age": 23423,
  },
]
`;
